### ISSUE TITLE
CI: new L7 field in CNP causes CRD validation failure in K8s 1.9 tests

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#5638]
I ran the K8s tests in a PR that did not touch CRD validation at all. All tests passed, except for the K8s 1.9 tests: 

Build link: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/887/#showFailuresLink
Test logs: 
[test_logs.zip](https://github.com/cilium/cilium/files/2399044/test_logs.zip)

Relevant output (from `cilium-test.log` in the above archive).
```
2018-09-19T22:00:58.067487895Z level=debug msg="CRD validation is different, updating it..." name=CiliumNetworkPolicy/v2 subsys=k8s
2018-09-19T22:00:58.083637896Z level=debug msg="Unable to update CRD validation" error="CustomResourceDefinition.apiextensions.k8s.io \"ciliumnetworkpolicies.cilium.io\" is invalid: [spec.validation.openAPIV3Schema.properties[EgressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[L7Rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[IngressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRule].properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRuleL7].additionalProperties: Forbidden: additionalProperties cannot be set to false]" name=CiliumNetworkPolicy/v2 subsys=k8s
2018-09-19T22:00:58.083677402Z level=error msg="Unable to update CRD" error="CustomResourceDefinition.apiextensions.k8s.io \"ciliumnetworkpolicies.cilium.io\" is invalid: [spec.validation.openAPIV3Schema.properties[EgressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[L7Rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[IngressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRule].properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRuleL7].additionalProperties: Forbidden: additionalProperties cannot be set to false]" name=CiliumNetworkPolicy/v2 subsys=k8s
2018-09-19T22:00:58.083808811Z level=fatal msg="Unable to establish connection to Kubernetes apiserver" error="Unable to create custom resource definition: CustomResourceDefinition.apiextensions.k8s.io \"ciliumnetworkpolicies.cilium.io\" is invalid: [spec.validation.openAPIV3Schema.properties[EgressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[L7Rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[spec].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[IngressRule].properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRule].properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[specs].items.properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[egress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[Rule].properties[ingress].items.properties[toPorts].items.properties[rules].properties[l7].items.additionalProperties: Forbidden: additionalProperties cannot be set to false, spec.validation.openAPIV3Schema.properties[PortRuleL7].additionalProperties: Forbidden: additionalProperties cannot be set to false]" subsys=daemon
```

I'll try running the tests again in my PR to see if this reproduces.

**Comment by tgraf** [comment]
Looks related to merging #5174

**Comment by ianvernon** [comment]
I wonder if certain restrictions in CRDs are not supported in K8s 1.9 vs other versions ? =/

**Comment by jrajahalme** [comment]
Need to figure out how to detect the cluster k8s version and add the validation of the generic L7 policies only if >= 1.11. Earlier versions do not support unspecified maps in the schema.

**Comment by tgraf** [comment]
@jrajahalme `func GetServerVersion() (ver *go_version.Version, err error) {` will help

**Comment by jrajahalme** [comment]
https://github.com/cilium/cilium/pull/5640

**Comment by jrajahalme** [comment]
Simply commenting put the offending part seems to work on most k8s versions, i tested this locally and it worked on k8s 1.10 and 1.11, but I did not get 1.9 running in minikube even without starting cilium at all.

**Comment by jrajahalme** [comment]
Fixed by #5640, clean CI run including k8s 1.9 .
